cabal-install/Distribution/Client/SetupWrapper.hs

```diff
@@ -304,7 +304,8 @@ internalSetupMethod verbosity options _ bt mkargs = do
   info verbosity $ "Using internal setup method with build-type " ++ show bt
                 ++ " and args:\n  " ++ show args
   inDir (useWorkingDir options) $
-    buildTypeAction bt args
+    withEnv "HASKELL_DIST_DIR" (useDistPref options) $
```

Comment:
If you're doing `./Setup test` the test script might care too. So really it needs to apply in all cases.